js/src/components/tree-select-control/index.js

```diff
@@ -100,6 +112,67 @@ const TreeSelectControl = ( {
 		setTreeVisible( false );
 	} );
 
+	/**
+	 * Perform the search query filter in the Tree options
+	 *
+	 * 1. Check if the search query is already cached and return it if so.
+	 * 2. Deep Copy the tree. Since we are going to modify his children and labels recursively.
+	 * 3. In case of filter, we apply the filter option function to the tree.
+	 * 4. In the filter function we also highlight the label with the matching letters
+	 * 5. Finally we set the cache with the obtained results and apply the filters
+	 *
+	 */
+	const queryOptions = useCallback( () => {
+		const cachedFilteredOptions = filteredOptionsCache.current[ filter ];
+
+		const highlightOptionLabel = ( optionLabel, matchPosition ) => {
+			const matchLength = matchPosition + filter.length;
+			return (
+				<span>
+					<span>{ optionLabel.substring( 0, matchPosition ) }</span>
+					<strong>
+						{ optionLabel.substring( matchPosition, matchLength ) }
+					</strong>
+					<span>{ optionLabel.substring( matchLength ) }</span>
+				</span>
+			);
+		};
+
+		const filterOption = ( option ) => {
+			if ( hasChildren( option ) ) {
+				option.children = option.children.filter( filterOption );
+				return !! option.children.length;
+			}
+
+			const match = option.label.toLowerCase().indexOf( filter );
+
+			if ( match >= 0 ) {
+				option.label = highlightOptionLabel( option.label, match );
+				return true;
+			}
+		};
+
+		if ( cachedFilteredOptions ) {
+			setFilteredOptions( cachedFilteredOptions );
+			return;
+		}
+
+		let filteredTreeOptions = cloneDeep( treeOptions );
+
+		if ( filter ) {
+			filteredTreeOptions = filteredTreeOptions.filter( filterOption );
+		}
+
+		filteredOptionsCache.current[ filter ] = filteredTreeOptions;
+		setFilteredOptions( filteredTreeOptions );
+	}, [ treeOptions, filter ] );
+
+	useEffect( () => {
+		queryOptions();
+	}, [ queryOptions ] );
```

Comment:
💅 
The `useCallback` at [line 125](https://github.com/woocommerce/google-listings-and-ads/blob/f222e095f1c3aa94d496283990b22f3621a31767/js/src/components/tree-select-control/index.js#L125)
```js
const queryOptions = useCallback( () => {
```
could be replaced with
```js
useEffect( () => {
```

Thus, the `useEffect` at lines 170-172 could be omitted. But if the descriptive of purpose is better in this way now, then it's good to keep it as-is.